### PR TITLE
fix: Dynamic adding Carousel data will automatically switch to the first page

### LINKED ISSUE #24879: Dynamic adding Carousel data will automatically switch to the first page
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/cool-leftpad-ru8qy?file=/index.js)

### Steps to reproduce
1. 点击切换到第二页；
2. 动态增加数据；

### What is expected?
不自动切换回第一页，即保持当前页不动

### What is actually happening?
每次添加数据后都会自动切换至第一页。

| Environment | Info |
|---|---|
| antd | 3.26.15 |
| React | 16.8 |
| System | window |
| Browser | chrome |

---
直接使用react-slick并无此问题
https://codesandbox.io/s/jinyong-ant-design-demo-3knzc
image

### PR DESCRIPTION
**Pull request by yoyo837**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Perfermance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
fix #24879

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     fix: Dynamic adding Carousel data will automatically switch to the first page      |
| 🇨🇳 Chinese |     修复：动态修改项的数量时如果当前索引仍然有效则保持，否则才重置到第一个项目      |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by yoyo837** [comment]
@gaoyiniy 这样改如何？

File: components/carousel/index.tsx [hunk 1]
```diff
@@ -49,9 +49,13 @@ export default class Carousel extends React.Component<CarouselProps, {}> {
...
-  componentDidUpdate(prevProps: CarouselProps) {
-    if (React.Children.count(this.props.children) !== React.Children.count(prevProps.children)) {
-      this.goTo(this.props.initialSlide || 0, false);
+  componentDidUpdate() {
+    const { children, initialSlide = 0 } = this.props;
+    const newLength = React.Children.count(children);
+    const current = this.innerSlider.state.currentSlide;
+    if (current > newLength - 1) {
+      const useIdx = initialSlide < newLength ? initialSlide : 0;
+      this.goTo(useIdx, false);
```
Comment (afc163) [inline components/carousel/index.tsx:58]: 把 componentDidUpdate 逻辑都移到 @ant-design/react-slick 里去吧。

Comment (yoyo837) [reply components/carousel/index.tsx:58]: OK，先关了。

Comment (yoyo837) [reply components/carousel/index.tsx:58]: 这里有个相关的逻辑了，是不是可以直接改这个？
https://github.com/ant-design/react-slick/blob/2f09a62846df9793157ac2a48429201dc4d6e60e/src/inner-slider.js#L153

**Pull request closed without merging**